[WFLY-10712] Observability: MicroProfile Metrics

jmesnil opened the pull request:
Add microprofile-metrics-smallrye extension to provide support for
MicroProfile Metrics 1.1
* integrate smallrye-metrics:1.1.2 that implements MicroProfile Metrics
  1.1
* provide the /metrics HTTP endpoint on the management interface
* The extension is added to standalone profiles only (not available in
  domain mode)

Dev analysis: wildfly/wildfly-proposals#147
JIRA: https://issues.jboss.org/browse/WFLY-10712

jmesnil committed:
[WFLY-10712] Observability: MicroProfile Metrics

Add microprofile-metrics-smallrye extension to provide support for
MicroProfile Metrics 1.1
* integrate smallrye-metrics:1.1.2 that implements MicroProfile Metrics
  1.1
* provide the /metrics HTTP endpoint on the management interface
* The extension is added to standalone profiles only (not available in
  domain mode)

Dev analysis: https://github.com/wildfly/wildfly-proposals/pull/147
JIRA: https://issues.jboss.org/browse/WFLY-10712

jmesnil committed:
Register JAXRS attributes as runtime read-only attribute

jmesnil commented:
@rsvoboda fiy I've squashed all the commits from the dev branch in a single commit

The community documentation is in progress, I'll add it a commit later.

jmesnil committed:
[WFLY-10712] Observability: MicroProfile Metrics

* Add documentation in the admin guide.

JIRA: https://issues.jboss.org/browse/WFLY-10712

bstansberry commented on microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MicroProfileMetricsSubsystemDefinition.java:
```diff
@@ -0,0 +1,79 @@
+    static final String METRICS_REGISTRATION_CAPABILITY = "org.wildlfy.extension.microprofile.metrics.wildfly-registrar";
+    static final RuntimeCapability<Void> METRICS_REGISTRATION_RUNTIME_CAPABILITY = RuntimeCapability.Builder.of(METRICS_REGISTRATION_CAPABILITY)
+            .build();
+    public static final ServiceName WILDFLY_REGISTRATION_SERVICE = ServiceName.parse(METRICS_REGISTRATION_CAPABILITY);
```
This should be obtained from METRICS_REGISTRATION_RUNTIME_CAPABILITY and not reparsed. Reparsing will result in extra strings in the heap. The MP Health subsystem has the same problem.

bstansberry commented on microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MicroProfileMetricsSubsystemRemove.java:
```diff
@@ -0,0 +1,42 @@
+
+        context.removeService(MetricsContextService.SERVICE_NAME);
+        context.removeService(MicroProfileMetricsSubsystemDefinition.WILDFLY_REGISTRATION_SERVICE);
+    }
```
I think ServiceRemoveStepHandler can do this.

Tangent: it might be nicer if ServiceRemoveStepHandler didn't need to be told what capabilities to remove in its constructor and could be set to just see what's registered with the MRR.

jmesnil replied:
I think I have fixed it but the `ServiceRemoveStepHandler(org.jboss.as.controller.AbstractAddStepHandler, org.jboss.as.controller.capability.RuntimeCapability...)` constructor I used is deprecated. I'm not sure why it is deprecated as my use case seems valid (2 capabilities are associated to the resource).

bstansberry replied:
@jmesnil The deprecated probably relates to my point about "it might be nicer if ServiceRemoveStepHandler didn't need to be told what capabilities to remove in its constructor and could be set to just see what's registered with the MRR."  A lot of the core handlers took RuntimeCapability... as a constructor param and used the data to record the capability. But then they were changed to get the capabilities from the MRR, making adding via the constructor an out-of-date code style. So all those constructor variants got deprecated. But ServiceRemoveStepHandler uses the capabilities for more than just recording them. So the deprecation there was premature.

bstansberry replied:
Oh, I see that my "it might be nicer" comment is wrong. Looking more closely ServiceRemoveStepHandler already does what I want. It gets the capabilities from the MRR if they were not provided via the constructor param.

So, I'm not sure what to say about the deprecation, other than perhaps this c'tor should be public now:

```
protected ServiceRemoveStepHandler(final AbstractAddStepHandler addOperation)
```

bstansberry commented on microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/_private/MicroProfileMetricsLogger.java:
```diff
@@ -0,0 +1,75 @@
+
+    @LogMessage(level = WARN)
+    @Message(id = 3, value = "Type %s is not supported for MicroProfile Metrics, the attribute %s on %s will not be registered.")
+    void unsupportedModelType(ModelType type, String attributeName, PathAddress address);
```
If this is called, either our software has a bug (metric with a bad type) or someone is deliberately trying to do something fancy that this subsystem can't handle. I don't think either should result in a log WARN, which is something an end user admin would be expected to react to.

jmesnil replied:
fixed by logging at debug level.

This warn was a reminder that some valid metrics (e.g. LIST of numerical values) are not supported by the extension but that is not something that should be reported to the end user/admin.

bstansberry replied:
👍

bstansberry commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/microprofile/metrics-smallrye/main/module.xml:
```diff
@@ -0,0 +1,53 @@
+        <module name="org.jboss.msc"/>
+        <module name="org.jboss.logging"/>
+        <module name="org.jboss.vfs"/>
+        <module name="org.jboss.as.weld" />
```
Why?

jmesnil replied:
dump copy/paste is not a good reason? ;)

bstansberry replied:
:-)

jmesnil commented on microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MetricsRegistrationService.java:
```diff
@@ -0,0 +1,315 @@
+
+        ServiceBuilder<?> serviceBuilder = context.getServiceTarget().addService(MicroProfileMetricsSubsystemDefinition.WILDFLY_REGISTRATION_SERVICE);
+        final ServiceName modelControllerName = Services.JBOSS_SERVER_CONTROLLER;
+        Supplier<ModelController> modelController = serviceBuilder.requires(modelControllerName);
```
fixed

jmesnil commented on microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MetricsRegistrationService.java:
```diff
@@ -0,0 +1,315 @@
+            return;
+        }
+
+        Set<String> attributeNames = managementResourceRegistration.getAttributeNames(address);
```
fixed

rsvoboda commented:
I played a bit with this at home yesterday before going to bed and I noticed issues in docs. I just created https://issues.jboss.org/browse/WFWIP-150.
There are some issues that need my attention for 7.2, not sure if I will be able to come back to MP Metric before FF of WF 15.

jmesnil committed:
Add batch job service to the DEPLOYMENT_COMPLETE_SERVICES list

This ensures that the deployment will be considered complete once Batch
jobs services have been properly installed.

jmesnil commented:
@jamezp could you please review 
https://github.com/wildfly/wildfly/pull/11830/commits/f89b5f5d059eefd91805eb78631aa67030d4318 and approve it?

jamezp commented on microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/_private/MicroProfileMetricsLogger.java:
```diff
@@ -0,0 +1,69 @@
+@MessageLogger(projectCode = "WFLYMETRICS", length = 4)
+public interface MicroProfileMetricsLogger extends BasicLogger {
+    /**
+     * A logger with the category {@code org.wildfly.extension.batch}.
```
Not a big deal, but we might want to change the category name here.

jamezp commented on microprofile/metrics-smallrye/src/main/resources/subsystem-templates/microprofile-metrics-smallrye.xml:
```diff
@@ -0,0 +1,10 @@
+<config>
+    <extension-module>org.wildfly.extension.microprofile.metrics-smallrye</extension-module>
+    <subsystem xmlns="urn:wildfly:microprofile-metrics-smallrye:1.0"
+               security-enabled="false"
```
Should this be `true`? The default for in the XSD and on the attribute is `true`.

jmesnil replied:
yes, the setting is correct.
By default, security is enabled. However for our standalone configuration where the management HTTP interface is split from the application HTTP interface we disable authentication.

kabir committed:
Merge branch 'master' into WFLY-10712_microprofile_metrics-smallrye
